Feat : Fcm 알림 전송 로직 구현 및 알림 목록 조회 API 추가

hyeonda02 opened the pull request:
### 👀 관련 이슈
#42 
생각보다 알림 목록 조회 로직이 복잡해서 변경된 코드가 많은 점, 죄송합니다.
앞으로는 이슈 단위를 더 작게하여 작업을 진행하도록 하겠습니다.

<!-- 관련 이슈를 적어주세요 -->

### ✨ 작업한 내용

<!-- 작업한 내용을 적어주세요 -->


### 1. FCM 등록 및 비활성화 API 추가 
알림 기능을 이용하기 위해 FCM Token을 저장하는 API와 로그아웃 시 이를 비활성화 하는 API를 추가하였습니다. 스웨거 테스트 진행한 결과와 기능에 대한 자세한 설명을 노션 명세서에 작성해두었습니다. 


### 2. 오전 9시마다 투두 알림을 보내는 스케줄러 추가 
알림 보내는 로직을 구현하기 위해 `FirebaseConfig`, `FcmNotificationSender`, `DailyTodoReminerScheduler` 구현하였습니다. 오전 9시마다 당일에 투두가 존재하는 사용자(active 상태의 fcm토큰)에게 알림을 보내도록 구현하였습니다. 노션에 오전 9시마다 전달되는 “오늘의 투두 알람” 에 알림 데이터 형식을 작성해 두었습니다.   
현재 실제 유효한 FDM Device Token이 없어서 FCM 전송 자체에 대한 테스트는 불가능했습니다.  대신 아래와 같은 방식으로 기능 흐름을 검정하였습니다


(1) 스케줄러 작동 확인  
- 1분 마다 실행되도록 구성  
- `[9am 스케줄러] 오늘의 투두 알림 메서드 호출` 로그 출력으로 정상 작동 확인 

(2) 일정 조회 로직 확인 
- SQL 로그를 통해 유저별로 일정 정상 조회 확인 

(3) 알림 페이로드 생성 확인 
-  `NotificationPayload` 객체 생성 직후 로그를 통해 일정 내용을 기반으로 정상 생성되었음을 확인 

(4) FCM 전송 로직 진입 여부 확인 
- User별로 Active 상태의 토큰을 이용해 FCM 알림 전송 기능 동작 확인 
- 실제 전송은 실패했지만 `"FCM 전송: Message 생성 완료"` 로그 출력 확인 - 가짜 토큰에 대해 `"The registration token is not a valid ~”` 에러 발생 → 전송 로직까지는 진입 완료

<img width="1124" height="169" alt="스크린샷 2025-07-22 오전 5 41 48" src="https://github.com/user-attachments/assets/81c1ad25-4613-40a9-adcf-42229bb25ffc" />

### **3. 알림 목록을 조회하는 API 추가(GET : `/api/notificatoins`)**
알림 엔티티의 구조를 수정하였습니다.
```
public class Notification extends BaseEntity {
  private Long id;
  private User user;
  private NotificationType type;
  private  String content;
  private Boolean isRead = false;
  private Long relatedId;  // 알림과 연관된 도메인의 식별자입니다. NotificationType에 따라 의미가 달라집니다.

}
```

User는 알림을 받는 사람을 의미합니다. 	
NotificationType은 알림에서 보여지는 문구, 알림 목록 화면에서 보여지는 문구, 알림 로직을 처리하기 위한 	 RelatedEntityType(enum)으로 구성되어 있습니다.   	

**알림 목록 조회 로직은 다음과 같습니다.** 	
틈 요청을 생성할때, 틈 응답이 생성 및 수정 될때, 틈 거절이 되었을떄, 팔로우 요청이 왔을떄 	
FcmNotificationSender.send()를 이용하여 알림을 보내고 Notification Entity를 생성합니다. 	
그리고 알림 목록을 조회하는 API에서 Notification Entity를 조회합니다.  	

FcmToken이 없기 때문에 현재 직접적인 테스트는 불가능한 상황입니다. 	
따라서 조회 로직이 정상적으로 동작하는지 Mock data를 넣어서 스웨거 테스트만 진행하였습니다. 
	 	 	
현재까지의 기획에서는 틈 요청 보낼때, 틈 응답(수락,거절,제안)보낼때, 스케줄 취소할때, 팔로우 했을떄 알림이 전송되는 것으로 알고 있는데 	해당 api에 대해 알림을 보내는 로직은 이후의 이슈에서 제가 진행하도록 하겠습니다.          
알림 관련 세팅과 목록 조회 로직을 다른 분들과 얘기를 나눈 후 진행한게 아니라서 이걸 설명해드리는 것도 시간이 걸릴거 같아서요..!



### 🌀 PR Point

<!-- 코드리뷰가 필요한 부분이 있다면 적어주세요 -->
알림 조회 로직에서 개선할 점이 있다면 편하게 말씀 주세요!

### 🍰 참고사항

<!-- 참고할 사항이 있다면 적어주세요 -->
서브모듈이 업데이트 되었습니다! 
그리고 CD.yml에서 Firebase 파일을 생성하는 단계를 추가해 두었는데, 잘 동작하지 않을 수도 있습니다.
만약 merge 된 후 서버에 문제가 생긴다면 바로 확인하도록 하겠습니다.

### 📷 스크린샷 또는 GIF

전부 다 포함하면 많을 것 같아서 일부만 첨부합니다..!

| 기능 | 스크린샷 |
| :--: | :------: |
|   알림 목록 조회 API - 틈 요청 전송   |  <img width="1294" height="395" alt="스크린샷 2025-07-23 오전 4 17 12" src="https://github.com/user-attachments/assets/7e356fa8-ce22-4d06-800f-9ec120b43954" /> |
|   알림 목록 조회 API - 팔로우 알림   |  <img width="1295" height="246" alt="스크린샷 2025-07-23 오전 4 17 22" src="https://github.com/user-attachments/assets/3062b761-f712-483a-960b-c6778748992d" /> |
|   알림 목록 조회 API - 요청 거절   |  <img width="1282" height="395" alt="스크린샷 2025-07-23 오전 4 16 29" src="https://github.com/user-attachments/assets/960bb48d-9a9b-49e3-bb74-221af6c25b69" /> |

hyeonda02 committed:
Feat : FCM 토큰 등록 및 비활성화 API 구현
- 다중 세션을 고려하여 FcmToken Entity 추가
- FcmToken -> User 단방향 @ManyToOne 연관관계 설계
- FCM 토큰 등록 API 구현
- FCM 토큰 비활성화 API 구현

hyeonda02 committed:
Feat : Fcm 관련 로직 및 스케줄러 추가
- FirebaseApp을 설정하기 위한 FirebaseConfig를 추가
- 알림을 보내기 위해 FcmNotificationSender 작성
- 알림 문자 템플릿을 위한 NotificationType 및 Payload 작성
- 오전 9시에 알림 보내는 TodoReminderScheduler 작성

hyeonda02 committed:
Feat : 알림 목록 조회 API 추가
- Prod 환경에서도 동작하도록 FirebaseConfig 수정
- Notification, NotificationType 엔티티 수정
- 알림 목록 조회 로직 추가

hyeonda02 committed:
Feat : NotificationService 로직에 틈 요청이 취소되었을 경우 처리 추가

hyeonda02 committed:
Refactor : FCM 및 알림 관련 API를 @CurrentUser 어노테이션을 통해 인증된 사용자 정보를 받아오도록 수정

hyeonda02 committed:
Refactor : Fcm 서비스 로직에 예외 처리 추가 및 FcmErrorStatus 코드 추가
- Service 로직에 예외 처리 코드 추가
- FcmSuccessStatus, FcmErrorStatus 상태코드 추가
- FcmTokenRequestDto에 @NotBlank 추가

hyeonda02 committed:
Refactor : ScheduleRepository JpaMethod 명 변경 및 스케줄러 알림 기능 테스트를 위해 로그 추가

Jinseong01 commented:
아직 정리 중이신 거 같은데 다 되시면 말씀해주세요!

hyeonda02 committed:
Refactor : 알림 응답 변환 시 관련 사용자 정보 매핑 로직 개선
- NotificationServiceImpl의 resolveRelatedUser 메서드 개선
- Notification Entity에 Not Null 조건 추가

hyeonda02 commented:
알림 기능 테스트 및 리팩토링 완료했습니다! 코드가 좀 많아서 리뷰 요청 드리는게 죄송스럽네요ㅠㅡ 다음부턴 이슈 단위를 좀 작게 해서 작업 진행하도록 하겠습니다.

hyeonda02 committed:
Refactor : NotificationConverter에서 S3Util 분리

Jinseong01 commented on src/main/java/umc/teumteum/server/domain/notification/service/NotificationServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+    return notifications.stream()
+        .map(n -> {
+          User friend = relatedUserMap.get(n.getId());
+          String profileImageUrl = s3Util.toPresignedUrl(friend.getProfileImageKey(), Duration.ofMinutes(30));
```
RelatedEntityType이 NONE인 경우에는 relatedUserMap에 { 알림 기본키 : null }의 형태로 들어가게 될 거 같은데
null 체크가 없어서 NullPointerException이 발생할 거 같습니다!

Jinseong01 commented on src/main/java/umc/teumteum/server/domain/notification/service/NotificationServiceImpl.java:
```diff
@@ -0,0 +1,71 @@
+
+      switch (type){
+        case TEUM_RESPONSE, SCHEDULE -> {
+          teumResponseRepository.findById(relatedId)
```
RelatedEntityType에 따라 Notification의 relatedId에

TEUM_RESPONSE, SCHEDULE -> 틈 응답 테이블 기본키
TEUM_REQUEST -> 틈 요청 테이블 기본키
FRIEND -> 친구 테이블 기본키
NONE -> ??
가 저장되고 알림 내용에 필요한 사용자의 데이터를 추가로 불러오는 식으로 작성된 거 같습니다.

1) 제가 이해한게 맞을까요??
2) NONE의 경우에는 연관된 엔티티가 없어서 다른 테이블로부터 정보를 추가 조회하지 않아도 된다는 것으로 이해했습니다. 그런데 Notification엔티티에는 relatedId가 nullable = false로 설정되어 있습니다. 확인 부탁드립니다!

hyeonda02 replied:
1. 넵 맞습니다! Notification의 content컬럼에는 NotificatoinType의 content 값을 저장할 계획입니다. 
사실 원래 9시에 보내는 투두 알림도 목록에서 보여져야 한다고 생각하고, 그 알림은 NONE enum을 사용하여 저장하기 위해 NONE을 설계했는데요. 그런데 나중에 보니 투두 알림은 저장할 필요가 없을 것 같아서 NONE enum 은 현재는 사용되지 않고 있습니다..! 오해를 불러올 수도 있으니 NONE 은 제거하도록 하겠습니다.

2. Notification 엔티티 컬럼을 보니 nullable=true로 했어야 했는데, 착각해서 Nullable =false로 한거 같습니다.... 새벽에 작업하다 보니 이런 멍청한 실수를;;;; 말씀 안해주셨으면 모르고 있었을 것 같습니다.. 꼼꼼하게 봐주셔서 감사합니다!

Jinseong01 replied:
9시 스케줄러에서 그날의 스케줄 보내는 알림은 DB 저장이 없다고 하셨는데
2번이 그래서 nullable =false로 설정하신 거일 수도 있을 거 같아요!

맞다면 기존 false 유지하고
default문으로 처리될만한 타입이 없다면 switch문 default분기에서 "처리 불가한 타입입니다"식의 예외 발생이 맞는거 같기도 합니다!! (그렇게 되면 위에서 말씀드린 NullPointerException도 발생 안 할 거 같습니다)

Jinseong01 commented on src/main/java/umc/teumteum/server/domain/notification/controller/NotificationController.java:
```diff
@@ -1,27 +1,37 @@
-    return null;
+  public ApiResponse<List<NotificationResponseDto>> getNotifications(
+      @CurrentUser @Parameter(hidden = true) User user){
+    List<NotificationResponseDto> notifications = notificationServiceImpl.getNotifications(user);
```
지금은 전체 목록 조회로 구현된 거 같아서 추후에 Slice 형식으로 리팩토링 진행해주시면 감사하겠습니다!

Jinseong01 commented on src/main/java/umc/teumteum/server/domain/notification/entity/Notification.java:
```diff
@@ -0,0 +1,52 @@
+
+  @Builder.Default
+  @Column(name = "is_read", nullable = false)
+  private Boolean isRead = false;
```
isRead는 알림 확인 여부를 나타내기 위한 필드 같습니다.
틈 응답에서는 상세조회 API를 통해 사용자가 읽었는지 여부를 변경할 수 있을 거 같은데요.
알림의 경우에도 여부를 변경할 수 있는 이벤트(?)가 있나요?? (누르면 관련된 화면 이동 등)

이런 이벤트가 있다면, 각각 알맞은 기존 API들 호출이 필요할 거 같은데 알림에 의한 호출인지에 따라서 isRead를 변경하는 로직이 필요하겠네요... 🤯

hyeonda02 replied:
ㅜㅠ 아 그럴수도 있겠군요.. 해당 알림을 누를 경우 화면 이동 같은 이벤트들이 있었던 것 같은데 그런 부분들도 생각해서 isRead 변경 로직을 추가해야될 것 같네요ㅠ
어차피 틈 요청을 보내는 API, 틈 응답을 보내는 API 등등 해당 API에 알람을 보내는 로직을 추가해야 하는데, 그 부분 작업하면서 변경하는 로직도 같이 추가해 보겠습니다! 꼼꼼한 리뷰 정말 감사합니다 👍🥺
일단

1. 엔티티 컬럼에 nullable 변경하기
2. RelatedEntityType 에서 NONE 제거하기
3. Slice 형식으로 변경하기
4. 충돌 해결하기

이렇게 진행한 이후, 다시 리뷰 요청 드리겠습니다!
isRead 변경 로직은 새로운 이슈 파고, 알림 보내는 로직 작업하면서 진행해도 될까요?

Jinseong01 replied:
> 맞다면 기존 false 유지하고
default문으로 처리될만한 타입이 없다면 switch문 default분기에서 "처리 불가한 타입입니다"식의 예외 발생이 맞는거 같기도 합니다!! (그렇게 되면 위에서 말씀드린 NullPointerException도 발생 안 할 거 같습니다)

시작하시기 전에 위에 작성한 내용 한번만 확인 부탁드릴게요!

Jinseong01 commented on src/main/java/umc/teumteum/server/global/scheduler/DailyTodoReminerScheduler.java:
```diff
@@ -0,0 +1,69 @@
+      List<FcmToken> tokens = fcmTokenRepository.findByUserAndIsActiveTrue(user);
+      for(FcmToken token : tokens) {
+        fcmNotificationSender.send(token.getToken(), payload);
+      }
```
9시마다 그날의 스케줄을 알림으로 전송하는 로직이 작성된 거 같은데요.
이를 DB에 저장하는 로직은 없는 거 같습니다! 확인 부탁드릴게요...

hyeonda02 replied:
엇 9시마다 전송되는 투두 알람은 화면 설계서 상, "알람 목록" 화면에서 표시하지 않아도 되서 따로 저장하지 않았습니다!
혹시 다른 화면에서 9시마다 전송되는 투두 알람을 조회해야 할까요?

Jinseong01 replied:
아...! 그렇군요..
그러면 알림 전송이 필요한 곳에서 sender() & DB 저장 이런 식이 되겠네요... 헷갈리게 해드려서 죄송합니다 🥲

hyeonda02 replied:
오우 괜찮습니당~ 혹시 제가 모르는 곳에서 투두 알림이 조회되는 곳이 있나 싶었어요!

Jinseong01 commented:
추가로 conflict도 확인 부탁드릴게요....

yangjiae12 committed:
Refactor: 틈 요청 조회 API의 페이지 요청을 1부터 시작하도록 변경

- page 파라미터를 1-based 기준으로 처리
- PageRequest 생성 시 page - 1 보정 적용

yangjiae12 committed:
Refactor: 팔로워 조회 API에 Slice 적용 및 페이지 1부터 시작으로 보정

- 기존 stream + subList 기반 페이징/정렬 로직 제거
- Spring Data JPA의 Pageable + Sort 조합 사용
- follower.nickname 기준 오름차순 정렬
- 클라이언트 페이지 번호는 1부터 시작하도록 처리

yangjiae12 committed:
Refactor: 팔로잉 조회 API에 Slice 적용 및 페이지 1부터 시작으로 보정

- 기존 stream + subList 기반 페이징/정렬 로직 제거
- Spring Data JPA의 Pageable + Sort 조합 사용
- isFavorite 기준 내림차순 -> nickname 기준 오름차순 정렬
- 클라이언트 페이지 번호는 1부터 시작하도록 처리

yangjiae12 committed:
Docs: Swagger 문서 개선을 위해 팔로워 조회 API에 @Parameter 설명 및 example 추가

yangjiae12 committed:
Refactor: 틈 요청 조회 API 페이지 번호 보정 책임을 서비스로 이동

- 컨트롤러에서는 page, size 그대로 전달
- 서비스 내부에서 1-based → 0-based 보정 처리
- DB 정렬 기준(readAt ASC, createdAt DESC) 포함

yangjiae12 committed:
Refactor: 팔로잉/팔로워 목록 응답을 PagingResponseDto로 감싸도록 변경

- 기존 Slice<T> 응답 구조에서 PagingResponseDto<T>로 변경하여 불필요한 데이터 제거
- hasNext 필드를 명시적으로 포함해 무한스크롤에 필요한 최소 정보만 제공

hyeonda02 committed:
Refactor : 알림 목록 조회 API에 NPE방지 로직 추가 및 Slice 페이징 적용

hyeonda02 commented:
회의때 얘기해주셨던 내용들 반영해서 관련 내용 찾아보고 리팩토링 하였습니다!

1. switch에서 -> 를 사용할 경우 break를 걸지 않아도 된다고 합니다. Java 14 이상부터의 문법이라네요..! 
2. switch에서 default 분기에서 그대로 예외를 던져버리면 `resolveRelatedUsers` 메서드가 그냥 중단되기 때문에 나머지 정상 알림들도 조회하지 못하는 상황이 생깁니다. 그래서 default 분기에서는 로그만 출력하도록 하였습니다. (어차피 여기서 예외 처리를 하려고 한들, `getNotifications` 메소드에서 notifications에 대해 스트림을 돌면서 알림 키 값으로 relatedUserMap을 조회하기 때문에 예외가 발생할 것 같더라구요.) 그래서 따로 private method인 `isValidNotification` 에서 유효한(연결된 유저가 있는)알림을 필터링하는 메소드를 추가하는 식으로 NPE를 막고자 하였습니다.  이런식으로 서비스 로직을 조금 변경하였고, Entity의 nullable 속성이나 RelationEntityType은 따로 변경하지 않았습니다. 
3. 알림 목록의 경우 무한 스크롤이여서 Slice 형식을 적용하였습니다. page 는 1로 시작하도록 설정하였습니다. Slice 형식을 적용하면서 Dto와 컨버터 코드를 수정하였습니다. 그리고 page와 size에 대한 유효성 검사는 컨트롤러 단에서 @RequestParam, @Validated 조합으로 하도록 설정했습니다.

hyeonda02 committed:
Chore : config 파일 수정
- ddl-auto를 update로 변경하였습니다.

hyeonda02 commented:
그리고 작업하면서 제가 작성한 알람 관련 로직 다시 봤는데요 ..
9시에 보내는 투두 스케줄러에서 sendDailyTodos() 메서드는 N+1 쿼리 문제가 발생할 수 있을 것 같다는 생각이 들었습니다. 해당 부분은 다음 이슈에서 리팩토링 하겠습니다!

hyeonda02 committed:
Merge branch 'develop' into feature/42-fcm

Jinseong01 commented:
작성하신 내용 확인해보니
사용자 관점에서는 예외 상황이 발생하더라도 나머지 정상 알림은 조회되는 것이 더 좋겠다는 생각이 드네요!

그리고 default분기와 NPE 관련해서는 default에서 예외를 발생시킨다면
> User friend = relatedUserMap.get(n.getId());
String profileImageUrl = s3Util.toPresignedUrl(friend.getProfileImageKey(), Duration.ofMinutes(30));

에서의 NPE 방지가 되지 않나라고 생각했는데 (해당 코드 실행 자체가 X)

> 어차피 여기서 예외 처리를 하려고 한들, getNotifications 메소드에서 notifications에 대해 스트림을 돌면서 알림 키 값으로 relatedUserMap을 조회하기 때문에 예외가 발생할 것 같더라구요.

작성하신 내용처럼 이렇게 되나요..? 어렵네요..

Jinseong01 commented:
N+1은 sendDailyTodos()에서 어느 조회 말씀하시는 건지 알 수 있을까요??

hyeonda02 commented:
1. 넵 저도 스위치에서 NPE 처리를 하면 될꺼라 생각했는데
<img width="879" height="267" alt="스크린샷 2025-07-26 오후 3 53 30" src="https://github.com/user-attachments/assets/9102cf96-9cb0-4ace-872b-4c12384fabd0" />

해당 부분에서` notificationList.stream()`을 실행할때, notification의 아이디를 이용해서 조회를 하기 때문에 NPE가 발생할 것 같습니다.. 그래서 별도로 검증하는 private 메서드를 만들고, `.filter `과정을 추가했어요! 

2. 
<img width="826" height="192" alt="스크린샷 2025-07-26 오후 3 56 39" src="https://github.com/user-attachments/assets/a16f48fb-1e8a-4cb9-8836-170503b65917" />

`sendDatilyTodos()`메서드에서` userRepository.findAll()` 이후에
각 유저마다 findByUserIdAndDate를 호출하고 있는데 해당 과정에서 유저 수만큼 쿼리가 반복 발생하니까 N+1 문제가 발생할 것 같다는 생각이 들었습니다.
유저 + 스케줄 데이터를 한번에 가져오는 방식으로 다른 이슈에서 리팩터링 해보겠습니다!

hyeonda02 commented:
> N+1은 sendDailyTodos()에서 어느 조회 말씀하시는 건지 알 수 있을까요??

Jinseong01 commented:
넵 확인했습니다! 수고하셨습니다!
